proselint/tools.py

```diff
@@ -1,501 +1,288 @@
-        for i, c in enumerate(text + "\n"):
-            if s == 0 and c in quotes and pc in seps:
-                start = i
-                s = 1
-                q = c
-            elif s == 1 and matching(c, q):
-                s = 2
-            elif s == 2:
-                if c in seps:
-                    ranges.append((start+1, i-1))
-                    start = None
-                    s = 0
-                else:
-                    s = 1
-            pc = c
-        return ranges
-
-    def position_in_ranges(ranges, position):
-        for start, end in ranges:
-            if start <= position < end:
-                return True
-        return False
-
-    return position_in_ranges(find_ranges(text), position)
-
-
-def detector_50_Cent(text):
-    """Determine whether 50 Cent is a topic."""
-    keywords = [
-        "50 Cent",
-        "rap",
-        "hip hop",
-        "Curtis James Jackson III",
-        "Curtis Jackson",
-        "Eminem",
-        "Dre",
-        "Get Rich or Die Tryin'",
-        "G-Unit",
-        "Street King Immortal",
-        "In da Club",
-        "Interscope",
-    ]
-    num_keywords = sum(word in text for word in keywords)
-    return ("50 Cent", float(num_keywords > 2))
+    if not isinstance(config, dict):
+        config = config_base.proselint_base
+    checks = get_checks(config)
 
+    results = {}
+    chars = 0
 
-def topics(text):
-    """Return a list of topics."""
-    detectors = [
-        detector_50_Cent
+    if len(paths) == 0:
+        # Use stdin if no paths were specified
+        log.info("No path specified -> will read from <stdin>")
+        results["<stdin>"] = lint(sys.stdin, config=config, checks=checks)
+    else:
+        # offer "outer" executor, to make multiprocessing more effective
+        exe = ProcessPoolExecutor() if config["parallelize"] else None
+        for file in filepaths:
+            log.debug("Analyzing '%s'", file.name)
+            try:
+                with file.open(encoding="utf-8", errors="replace") as _fh:
+                    content = _fh.read()
+            except Exception:
+                log.exception("[LintPath] Error opening '%s' -> will skip", file.name)
+                continue
+            results[file] = lint(content, config, checks, file.as_posix(), _exe=exe)
+            chars += len(content)
+
+    # fetch result from futures, if needed
+    for _file, _errors in results.items():
+        if len(_errors) > 0 and isinstance(_errors[0], Future):
+            _errors = [_e for _ft in _errors for _e in _ft.result()]
+            _errors = sorted(
+                _errors[: config["max_errors"]],
+                key=lambda e: (e[2], e[3]),
+            )
+            # memoizer could also iterate, fetch results and sort
+            memoize_future(_errors, _file.as_posix())
+            # write back data so no futures are returned
+            results[_file] = _errors
+
+    # bad style ... but
+    global last_char_count
```

Comment:
I would like to avoid globals. What is this used for?

Reply:
me too, but printing out some stats at the end is compelling. didn't want to clutter the lint()-return